content/2020-05-22-book-review-writing-an-interpreter-in-go/index.md

```diff
@@ -0,0 +1,581 @@
+
+#### prefix vs infix operators
+
+Take a look at this line: `-5 - 10`. Can you guess which `-` is a prefix and which is an infix?
+
+Correct! The `-` in front of the `5` is a prefix operator because it's next to the integer and denotes negativity. The `-` in between the two integers is the infix operator and denotes subtraction.
+
+#### binary operators
+
+The `+`, `-`, `*` and `/` are all types of infix operators, or binary operators, meaning they operate with two operands.
+
+#### Vaughan Pratt
+
+Professor at Stanford born in 1944. You can read more about him on [Wikipedia](https://en.wikipedia.org/wiki/Vaughan_Pratt).
+
+His parsing method is explained in his paper "Top Down Operator Precedence" which was published in 1973 (gasp!) and can be read [here](https://dl.acm.org/doi/10.1145/512927.512931).
+
+As a reminder to myself, `prefixParseFns` is used in the book and is the same as "nuds" for "null denotations" by Pratt. `infixParseFns` is equivalent to "leds" or "left denotations" by Pratt.
+
+#### terminology
+
+Thorsten points out some useful terms on page 63:
+
+- **prefix operator** - in front of operand
+- **postfix operator** - after operand like `foobar++`
+- **infix operator** - in between operands
+- **binary expressions** - the operator has two operands like `5 * 8`
+- **operator precedence** - also called "order of operations" (ring a bell?)
+
+#### IntegerLiteral
+
+I wrote a note to show an example of what an integer literal data structure looks like in our language:
+
+```go
+type IntegerLiteral struct {
+  Token token.Token
+  Value int64
+}
+```
+
+As you can see, we store the token type and the actual value. For example, `5` would look like this:
+
+```go
+{
+  Token: token.INT, // integer
+  Value: "5"
+}
+```
+
+Pretty cool, huh?
+
+Sidenote: I love that he abbreviates literal to `lit` in our helper functions. I laugh because it reminds me of the colloquial saying, "that's so [lit](https://www.urbandictionary.com/define.php?term=lit)!"
+
+#### prefixParseFns and infixParseFns
+
+I found this part of the book a bit confusing. I'm struggling to even explain what I don't know...If I have the integer `5`, why do I call `prefixParseFns`? Is it because I need to check if `5` has a prefix such as a minus sign "-"? Ah...now that I'm writing this, I think that's it. Because as a human, I can look at `-5` and say, "Yes, that has a prefix we need to parse." But as a machine, I don' know that from looking at the token alone because these tokens `5` and `-5` are the same token type but have different values.
```

Comment:
You got it. The parser always needs to start parsing what he expects to be there. That's why where start calling `parseStatement` when kicking off the parser, since that _should_ be in the source code.

There's a `t` missing in `don't`:

```suggestion
I found this part of the book a bit confusing. I'm struggling to even explain what I don't know...If I have the integer `5`, why do I call `prefixParseFns`? Is it because I need to check if `5` has a prefix such as a minus sign "-"? Ah...now that I'm writing this, I think that's it. Because as a human, I can look at `-5` and say, "Yes, that has a prefix we need to parse." But as a machine, I don't know that from looking at the token alone because these tokens `5` and `-5` are the same token type but have different values.
```

Reply:
Ahhh! Okay, that makes sense! Thanks for validating that!